src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1173 @@
+    pub fn new(rng: &mut impl Rng256) -> PinProtocolV1 {
+        let key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+        let pin_uv_auth_token = rng.gen_uniform_u8x32();
+        PinProtocolV1 {
+            key_agreement_key,
+            pin_uv_auth_token,
+            consecutive_pin_mismatches: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        }
+    }
+
+    /// Decrypts the encrypted pin_hash and compares it to the stored pin_hash.
+    /// Resets or decreases the PIN retries, depending on success or failure.
+    fn verify_pin_hash_enc(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        aes_dec_key: &crypto::aes256::DecryptionKey,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        match persistent_store.pin_hash() {
+            Some(pin_hash) => {
+                if self.consecutive_pin_mismatches >= 3 {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_BLOCKED);
+                }
+                // We need to copy the pin hash, because decrementing the pin retries below mutably
+                // borrows from the same persistent_store reference, and therefore invalidates the
+                // pin_hash reference.
+                let pin_hash = pin_hash.clone();
+                persistent_store.decr_pin_retries();
+                if pin_hash_enc.len() != PIN_AUTH_LENGTH {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_PIN_INVALID);
+                }
+
+                let iv = [0u8; 16];
+                let mut blocks = [[0u8; 16]; 1];
+                blocks[0].copy_from_slice(&pin_hash_enc[0..PIN_AUTH_LENGTH]);
+                cbc_decrypt(aes_dec_key, iv, &mut blocks);
+
+                if !bool::from(pin_hash.ct_eq(&blocks[0])) {
+                    self.key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+                    if persistent_store.pin_retries() == 0 {
+                        return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+                    }
+                    self.consecutive_pin_mismatches += 1;
+                    if self.consecutive_pin_mismatches >= 3 {
+                        return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_BLOCKED);
+                    }
+                    return Err(Ctap2StatusCode::CTAP2_ERR_PIN_INVALID);
+                }
+            }
+            // This status code is not explicitly mentioned in the specification.
+            None => return Err(Ctap2StatusCode::CTAP2_ERR_PIN_REQUIRED),
+        }
+        persistent_store.reset_pin_retries();
+        self.consecutive_pin_mismatches = 0;
+        Ok(())
+    }
+
+    /// Uses the self-owned and passed halves of the key agreement to generate the
+    /// shared secret for checking pin_auth and generating a decryption key.
+    fn create_decryption_key_verified(
```

Comment:
Nit:
- `exchange_decryption_key` would be clearer and sufficient IMO.
- It's not clear what `message` contains in this context. Could be worth documenting or using a more meaningful name.

Otherwise looks good.

Reply:
Done